File: Labs/Lab5/Solution.java
public boolean isValidBST(TreeNode root) {
    return isValidBST(root, null);
}

private boolean isValidBST(TreeNode node, TreeNode prev) {
    if (node == null) return true;

    if (!isValidBST(node.left, prev)) return false;

    // Check the current node's value
    if (prev != null && node.val <= prev.val) return false;
    prev = node;

    return isValidBST(node.right, prev);
}

File: Labs/Lab5/Solution.py
def isValidBST(self, root):
    return self.is_valid_bst(root, None)

def is_valid_bst(self, node, prev):
    if node is None:
        return True

    if not self.is_valid_bst(node.left, prev):
        return False

    # Check the current node's value
    if prev is not None and node.val <= prev.val:
        return False
    prev = node

    return self.is_valid_bst(node.right, prev)
